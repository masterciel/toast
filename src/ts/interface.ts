import _Vue from "vue";
import ToastContainer from "../components/VtToastContainer.vue";
import events from "./events";
import { TYPE, EVENTS } from "./constants";
import { getId } from "./utils";
import { ToastContent, ToastOptions, ToastID, PluginOptions } from "../types";

const ToastInterface = (Vue: typeof _Vue, globalOptions?: PluginOptions) => {
  new (Vue.extend(ToastContainer))({
    el: document.createElement("div"),
    propsData: globalOptions
  });
  /**
   * Display a toast
   */
  const toast = (content: ToastContent, options?: ToastOptions): ToastID => {
    const props = Object.assign(
      {},
      { id: getId(), type: TYPE.DEFAULT },
      options,
      { content }
    );
    events.$emit(EVENTS.ADD, props);
    return props.id;
  };
  /**
   * Clear all toasts
   */
  toast.clear = () => events.$emit(EVENTS.CLEAR);
  /**
   * Update Plugin Defaults
   */
  toast.updateDefaults = (update: PluginOptions) => {
    events.$emit(EVENTS.UPDATE_DEFAULTS, update);
  };
  /**
   * Dismiss toast specified by an id
   */
  toast.dismiss = (id: ToastID) => {
    events.$emit(EVENTS.DISMISS, id);
  };
  /**
   * Update Toast
   */
  toast.update = (
    id: ToastID,
    { content, options }: { content?: ToastContent; options?: ToastOptions },
    create = false
  ) => {
    events.$emit(EVENTS.UPDATE, {
      id,
      options: Object.assign({}, options, { content }),
      create
    });
  };
  /**
   * Display a success toast
   */
  toast.success = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.SUCCESS }
  ) => toast(content, Object.assign({}, options, { type: TYPE.SUCCESS }));

  /**
   * Display an info toast
   */
  toast.info = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.INFO }
  ) => toast(content, Object.assign({}, options, { type: TYPE.INFO }));

  /**
   * Display an error toast
   */
  toast.error = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.ERROR }
  ) => toast(content, Object.assign({}, options, { type: TYPE.ERROR }));

  /**
   * Display a warning toast
   */
  toast.warning = (
    content: ToastContent,
    options?: ToastOptions & { type?: TYPE.WARNING }
  ) => toast(content, Object.assign({}, options, { type: TYPE.WARNING }));

  return toast;
};

export default ToastInterface;
