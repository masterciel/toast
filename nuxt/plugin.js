import Vue from "vue";
import Toast from "vue-toastification";

Vue.use(Toast, <%= serialize(options.pluginOptions) %>)

export default function (ctx, inject) {
  inject('toast', Vue.$toast)
}
