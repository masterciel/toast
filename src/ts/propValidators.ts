import { PropType } from "vue";
import {
  ToastContent,
  CommonOptions,
  PluginOptions,
  ToastID,
  ToastOptions
} from "../types";
import { TYPE, POSITION, VT_NAMESPACE } from "./constants";

const COMMON = {
  type: {
    type: String as PropType<TYPE>,
    default: TYPE.DEFAULT
  },
  classNames: {
    type: [String, Array] as PropType<string | string[]>,
    default: () => []
  },
  trueBoolean: {
    type: Boolean,
    default: true
  }
};

const ICON = {
  type: COMMON.type,
  customIcon: {
    type: [String, Boolean, Object, Function] as PropType<
      NonNullable<CommonOptions["icon"]>
    >,
    default: true
  }
};

const CLOSE_BUTTON = {
  component: {
    type: [String, Object, Function, Boolean] as PropType<
      NonNullable<CommonOptions["closeButton"]>
    >,
    default: "button"
  },
  classNames: COMMON.classNames,
  showOnHover: Boolean,
  ariaLabel: {
    type: String,
    default: "close"
  }
};

const PROGRESS_BAR = {
  timeout: {
    type: [Number, Boolean] as PropType<number | false>,
    default: 5000
  },
  hideProgressBar: Boolean,
  isRunning: Boolean
};

const TRANSITION = {
  transition: {
    type: [Object, String] as PropType<
      NonNullable<PluginOptions["transition"]>
    >,
    default: `${VT_NAMESPACE}__bounce`
  },
  transitionDuration: {
    type: [Number, Object] as PropType<
      NonNullable<PluginOptions["transitionDuration"]>
    >,
    default: 750
  }
};

const CORE_TOAST = {
  position: {
    type: String as PropType<POSITION>,
    default: POSITION.TOP_RIGHT
  },
  draggable: COMMON.trueBoolean,
  draggablePercent: {
    type: Number,
    default: 0.6
  },
  pauseOnFocusLoss: COMMON.trueBoolean,
  pauseOnHover: COMMON.trueBoolean,
  closeOnClick: COMMON.trueBoolean,
  timeout: PROGRESS_BAR.timeout,
  hideProgressBar: PROGRESS_BAR.hideProgressBar,
  toastClassName: COMMON.classNames,
  bodyClassName: COMMON.classNames,
  icon: ICON.customIcon,
  closeButton: CLOSE_BUTTON.component,
  closeButtonClassName: CLOSE_BUTTON.classNames,
  showCloseButtonOnHover: CLOSE_BUTTON.showOnHover,
  accessibility: {
    type: Object as PropType<PluginOptions["accessibility"]>,
    default: () => ({
      toastRole: "alert",
      closeButtonLabel: "close"
    })
  },
  rtl: Boolean
};

const TOAST = {
  id: {
    type: [String, Number] as PropType<ToastID>,
    required: true
  },
  type: COMMON.type,
  content: {
    type: [String, Object, Function] as PropType<ToastContent>,
    required: true
  },
  onClick: Function as PropType<NonNullable<ToastOptions["onClick"]>>,
  onClose: Function as PropType<NonNullable<ToastOptions["onClose"]>>
};

const CONTAINER = {
  newestOnTop: COMMON.trueBoolean,
  maxToasts: {
    type: Number,
    default: 20
  },
  container: {
    type: (Element as unknown) as PropType<
      NonNullable<PluginOptions["container"]>
    >,
    default: () => document.body
  },
  filterBeforeCreate: Function as PropType<
    NonNullable<PluginOptions["filterBeforeCreate"]>
  >,
  filterToasts: Function as PropType<
    NonNullable<PluginOptions["filterToasts"]>
  >,
  containerClassName: COMMON.classNames
};

export default {
  CORE_TOAST,
  TOAST,
  CONTAINER,
  PROGRESS_BAR,
  ICON,
  TRANSITION,
  CLOSE_BUTTON
};
