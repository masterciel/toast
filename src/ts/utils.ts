import { Component, defineComponent } from "vue"
import type {
  ToastComponent,
  ToastContent,
  RenderableToastContent,
} from "../types"

interface DictionaryLike {
  [index: string]: unknown
}

// eslint-disable-next-line @typescript-eslint/ban-types
const isFunction = (value: unknown): value is Function =>
  typeof value === "function"

const isString = (value: unknown): value is string => typeof value === "string"

const isNonEmptyString = (value: unknown): value is string =>
  isString(value) && value.trim().length > 0

const isNumber = (value: unknown): value is number => typeof value === "number"

const isUndefined = (value: unknown): value is undefined =>
  typeof value === "undefined"

const isObject = (value: unknown): value is DictionaryLike =>
  typeof value === "object" && value !== null

const isJSX = (obj: unknown): obj is JSX.Element =>
  hasProp(obj, "tag") && isNonEmptyString(obj.tag)

const isTouchEvent = (event: Event): event is TouchEvent =>
  window.TouchEvent && event instanceof TouchEvent

const isToastComponent = (obj: unknown): obj is ToastComponent =>
  hasProp(obj, "component") && isToastContent(obj.component)

const isVueComponent = (c: unknown): c is Component => {
  if (isFunction(c)) {
    return true
  }
  return isObject(c)
}

const isToastContent = (obj: unknown): obj is ToastContent =>
  // Ignore undefined
  !isUndefined(obj) &&
  // Is a string
  (isString(obj) ||
    // Regular Vue component
    isVueComponent(obj) ||
    // Nested object
    isToastComponent(obj))

const isDOMRect = (obj: unknown): obj is DOMRect =>
  isObject(obj) &&
  ["height", "width", "right", "left", "top", "bottom"].every(p =>
    isNumber(obj[p])
  )

const hasProp = <O extends unknown, K extends PropertyKey>(
  obj: O,
  propKey: K
): obj is O & { [key in K]: unknown } =>
  Object.prototype.hasOwnProperty.call(obj, propKey)

/**
 * ID generator
 */
const getId = (i => () => i++)(0)

function getX(event: MouseEvent | TouchEvent) {
  return isTouchEvent(event) ? event.targetTouches[0].clientX : event.clientX
}

function getY(event: MouseEvent | TouchEvent) {
  return isTouchEvent(event) ? event.targetTouches[0].clientY : event.clientY
}

const removeElement = (el: Element) => {
  if (!isUndefined(el.remove)) {
    el.remove()
  } else if (el.parentNode) {
    el.parentNode.removeChild(el)
  }
}

const getVueComponentFromObj = (obj: ToastContent): RenderableToastContent => {
  if (isToastComponent(obj)) {
    // Recurse if component prop
    return getVueComponentFromObj(obj.component)
  }
  if (isJSX(obj)) {
    // Create render function for JSX
    return defineComponent({
      render() {
        return obj
      },
    })
  }
  // Return the actual object if regular vue component
  return obj
}

export {
  getId,
  getX,
  getY,
  removeElement,
  isString,
  isNonEmptyString,
  isToastContent,
  getVueComponentFromObj,
  hasProp,
  isUndefined,
  isDOMRect,
  isFunction,
}
