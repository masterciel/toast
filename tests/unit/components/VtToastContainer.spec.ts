import { loadPlugin } from "../../utils/plugin";
import {
  ToastOptionsAndRequiredContent,
  PluginOptions,
  ToastID,
  ToastOptionsAndContent
} from "@/types";
import { POSITION, VT_NAMESPACE } from "@/ts/constants";

describe("VtToastContainer", () => {
  it("snapshots with default value", () => {
    const { containerWrapper } = loadPlugin();
    expect(containerWrapper.element).toMatchSnapshot();
  });
  it("snapshots with classes", () => {
    const { containerWrapper } = loadPlugin({ containerClassName: "myclass" });
    expect(containerWrapper.element).toMatchSnapshot();
  });
  describe("setup", () => {
    it("removes element and reassigns", () => {
      const { containerWrapper } = loadPlugin();
      const container = document.createElement("div");
      const vm = (containerWrapper.vm as unknown) as {
        setup(container: HTMLElement): void;
      };
      expect(containerWrapper.element.parentElement).not.toBe(container);
      vm.setup(container);
      expect(containerWrapper.element.parentElement).toBe(container);
    });
  });
  describe("setToast", () => {
    it("sets toast with id", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        setToast(props: ToastOptionsAndRequiredContent): void;
        toasts: { [toastID: string]: ToastOptionsAndRequiredContent };
      };
      expect(vm.toasts).toEqual({});
      const toast: ToastOptionsAndRequiredContent = {
        content: "abc",
        id: "id"
      };
      vm.setToast(toast);
      expect(vm.toasts).toEqual({ id: toast });
    });
    it("ignores toast without id", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        setToast(props: ToastOptionsAndRequiredContent): void;
        toasts: { [toastID: string]: ToastOptionsAndRequiredContent };
      };
      expect(vm.toasts).toEqual({});
      const toast: ToastOptionsAndRequiredContent = { content: "abc" };
      vm.setToast(toast);
      expect(vm.toasts).toEqual({});
    });
  });
  describe("addToast", () => {
    it("uses default values if nothing was provided", () => {
      const filterBeforeCreate = jest.fn(toast => toast);
      const { containerWrapper } = loadPlugin({ filterBeforeCreate });
      const vm = (containerWrapper.vm as unknown) as {
        addToast(params: ToastOptionsAndRequiredContent): void;
        defaults: PluginOptions;
      };
      const toast: ToastOptionsAndRequiredContent = { content: "abc" };
      expect(filterBeforeCreate).not.toHaveBeenCalled();
      vm.addToast(toast);
      expect(filterBeforeCreate).toHaveBeenCalledWith(
        { ...vm.defaults, ...toast },
        []
      );
    });
    it("merges default with params", () => {
      const filterBeforeCreate = jest.fn(toast => toast);
      const { containerWrapper } = loadPlugin({ filterBeforeCreate });
      const vm = (containerWrapper.vm as unknown) as {
        addToast(params: ToastOptionsAndRequiredContent): void;
        defaults: PluginOptions;
      };
      const toast: ToastOptionsAndRequiredContent = {
        content: "abc",
        timeout: 1000,
        closeButton: false
      };
      expect(filterBeforeCreate).not.toHaveBeenCalled();
      vm.addToast(toast);
      expect(filterBeforeCreate).toHaveBeenCalledWith(
        { ...vm.defaults, ...toast },
        []
      );
    });
    it("uses default filterBeforeCreate if defined", () => {
      const filterBeforeCreate = jest.fn(toast => toast);
      const { containerWrapper } = loadPlugin({ filterBeforeCreate });
      const vm = (containerWrapper.vm as unknown) as {
        addToast(params: ToastOptionsAndRequiredContent): void;
      };
      const toast: ToastOptionsAndRequiredContent = { content: "abc" };
      expect(filterBeforeCreate).not.toHaveBeenCalled();
      vm.addToast(toast);
      expect(filterBeforeCreate).toHaveBeenCalled();
    });
    it("uses fallback filterBeforeCreate if not defined", () => {
      const filterBeforeCreate = jest.fn((): false => false);
      const { containerWrapper, localVue } = loadPlugin({ filterBeforeCreate });
      localVue.$toast.updateDefaults({ filterBeforeCreate: undefined });
      const vm = (containerWrapper.vm as unknown) as {
        setToast(params: ToastOptionsAndRequiredContent): void;
        addToast(params: ToastOptionsAndRequiredContent): void;
        defaults: PluginOptions;
      };
      const spySetToast = jest.spyOn(vm, "setToast");
      const toast: ToastOptionsAndRequiredContent = { content: "abc" };
      expect(filterBeforeCreate).not.toHaveBeenCalled();
      expect(spySetToast).not.toHaveBeenCalled();
      vm.addToast(toast);
      expect(filterBeforeCreate).not.toHaveBeenCalled();
      expect(spySetToast).toHaveBeenCalledWith({ ...vm.defaults, ...toast });
    });
    it("set toast if passes filterBeforeCreate", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        setToast(params: ToastOptionsAndRequiredContent): void;
        addToast(params: ToastOptionsAndRequiredContent): void;
        defaults: PluginOptions;
      };
      const spySetToast = jest.spyOn(vm, "setToast");
      const toast: ToastOptionsAndRequiredContent = { content: "abc" };
      expect(spySetToast).not.toHaveBeenCalled();
      vm.addToast(toast);
      expect(spySetToast).toHaveBeenCalledWith({ ...vm.defaults, ...toast });
    });
    it("does not set toast if fails filterBeforeCreate", () => {
      const filterBeforeCreate = jest.fn((): false => false);
      const { containerWrapper } = loadPlugin({ filterBeforeCreate });
      const vm = (containerWrapper.vm as unknown) as {
        setToast(params: ToastOptionsAndRequiredContent): void;
        addToast(params: ToastOptionsAndRequiredContent): void;
      };
      const spySetToast = jest.spyOn(vm, "setToast");
      const toast: ToastOptionsAndRequiredContent = { content: "abc" };
      expect(spySetToast).not.toHaveBeenCalled();
      vm.addToast(toast);
      expect(spySetToast).not.toHaveBeenCalled();
    });
  });
  describe("dismissToast", () => {
    it("dismisses toast", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        dismissToast(id: ToastID): void;
        setToast(params: ToastOptionsAndRequiredContent): void;
      };
      const spyOnDelete = jest.spyOn(containerWrapper.vm, "$delete");
      vm.setToast({ id: 10, content: "content" });
      expect(spyOnDelete).not.toHaveBeenCalled();
      vm.dismissToast(10);
      expect(spyOnDelete).toHaveBeenCalledWith(expect.anything(), 10);
    });
    it("calls onClose if set", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        dismissToast(id: ToastID): void;
        setToast(params: ToastOptionsAndRequiredContent): void;
      };
      const spyOnDelete = jest.spyOn(containerWrapper.vm, "$delete");
      const onClose = jest.fn();
      vm.setToast({ id: 10, content: "content", onClose });
      expect(spyOnDelete).not.toHaveBeenCalled();
      expect(onClose).not.toHaveBeenCalled();
      vm.dismissToast(10);
      expect(onClose).toHaveBeenCalled();
      expect(spyOnDelete).toHaveBeenCalledWith(expect.anything(), 10);
    });
  });
  describe("clearToasts", () => {
    it("clears toasts", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        dismissToast(id: ToastID): void;
        clearToasts(): void;
        setToast(params: ToastOptionsAndRequiredContent): void;
      };
      const spyOnDismiss = jest.spyOn(containerWrapper.vm, "$delete");
      vm.setToast({ id: 1, content: "content1" });
      vm.setToast({ id: 2, content: "content2" });
      vm.setToast({ id: 3, content: "content3" });
      expect(spyOnDismiss).not.toHaveBeenCalled();
      vm.clearToasts();
      expect(spyOnDismiss).toHaveBeenCalledTimes(3);
      expect(spyOnDismiss).toHaveBeenCalledWith(expect.anything(), "1");
      expect(spyOnDismiss).toHaveBeenCalledWith(expect.anything(), "2");
      expect(spyOnDismiss).toHaveBeenCalledWith(expect.anything(), "3");
    });
  });
  describe("getPositionToasts", () => {
    it("gets toast from position", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        getPositionToasts(position: POSITION): ToastOptionsAndRequiredContent[];
        setToast(params: ToastOptionsAndRequiredContent): void;
      };
      vm.setToast({
        id: 1,
        content: "content1",
        position: POSITION.TOP_CENTER
      });
      vm.setToast({ id: 2, content: "content1", position: POSITION.TOP_RIGHT });
      const topCenterIds = vm
        .getPositionToasts(POSITION.TOP_CENTER)
        .map(t => t.id);
      const topRightIds = vm
        .getPositionToasts(POSITION.TOP_RIGHT)
        .map(t => t.id);
      const topLeftIds = vm.getPositionToasts(POSITION.TOP_LEFT).map(t => t.id);
      expect(topCenterIds).toEqual([1]);
      expect(topRightIds).toEqual([2]);
      expect(topLeftIds).toEqual([]);
    });
    it("regular ordering if newestOnTop is false", () => {
      const { containerWrapper: containerWrapper1 } = loadPlugin({
        newestOnTop: false
      });
      const vm = (containerWrapper1.vm as unknown) as {
        getPositionToasts(position: POSITION): ToastOptionsAndRequiredContent[];
        setToast(params: ToastOptionsAndRequiredContent): void;
      };
      vm.setToast({ id: 1, content: "content1", position: POSITION.TOP_RIGHT });
      vm.setToast({ id: 2, content: "content2", position: POSITION.TOP_RIGHT });
      vm.setToast({ id: 3, content: "content3", position: POSITION.TOP_RIGHT });
      const topRightIds = vm
        .getPositionToasts(POSITION.TOP_RIGHT)
        .map(t => t.id);
      expect(topRightIds).toEqual([1, 2, 3]);
    });
    it("reverse ordering if newestOnTop is true", () => {
      const { containerWrapper: containerWrapper1 } = loadPlugin({
        newestOnTop: true
      });
      const vm = (containerWrapper1.vm as unknown) as {
        getPositionToasts(position: POSITION): ToastOptionsAndRequiredContent[];
        setToast(params: ToastOptionsAndRequiredContent): void;
      };
      vm.setToast({ id: 1, content: "content1", position: POSITION.TOP_RIGHT });
      vm.setToast({ id: 2, content: "content2", position: POSITION.TOP_RIGHT });
      vm.setToast({ id: 3, content: "content3", position: POSITION.TOP_RIGHT });
      const topRightIds = vm
        .getPositionToasts(POSITION.TOP_RIGHT)
        .map(t => t.id);
      expect(topRightIds).toEqual([3, 2, 1]);
    });
  });
  describe("updateDefaults", () => {
    it("updates defaults", () => {
      const { containerWrapper } = loadPlugin({ timeout: 1000 });
      const vm = (containerWrapper.vm as unknown) as {
        updateDefaults(update: PluginOptions): void;
        defaults: PluginOptions;
      };
      expect(vm.defaults.timeout).toBe(1000);
      vm.updateDefaults({ timeout: 5000 });
      expect(vm.defaults.timeout).toBe(5000);
    });
    it("calls setup if container is present", () => {
      const { containerWrapper } = loadPlugin({ timeout: 1000 });
      const vm = (containerWrapper.vm as unknown) as {
        updateDefaults(update: PluginOptions): void;
        defaults: PluginOptions;
        setup(container: HTMLElement): void;
      };
      const spySetup = jest.spyOn(vm, "setup");
      const container = document.createElement("div");
      expect(vm.defaults.container).not.toBe(container);
      expect(spySetup).not.toHaveBeenCalled();
      vm.updateDefaults({ container });
      expect(spySetup).toHaveBeenCalledWith(container);
      expect(vm.defaults.container).toBe(container);
    });
  });
  describe("updateToast", () => {
    it("updates existing toast", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        updateToast({
          id,
          options,
          create
        }: {
          id: ToastID;
          options: ToastOptionsAndContent;
          create: boolean;
        }): void;
        setToast(params: ToastOptionsAndRequiredContent): void;
        toasts: { [toastID: string]: ToastOptionsAndRequiredContent };
      };
      const toast: ToastOptionsAndRequiredContent = {
        id: "id1",
        content: "content"
      };
      vm.setToast(toast);
      expect(vm.toasts["id1"].content).toBe("content");
      vm.updateToast({
        id: "id1",
        options: { content: "other" },
        create: false
      });
      expect(vm.toasts["id1"].content).toBe("other");
    });
    it("increases timeout if it is the same", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        updateToast({
          id,
          options,
          create
        }: {
          id: ToastID;
          options: ToastOptionsAndContent;
          create: boolean;
        }): void;
        setToast(params: ToastOptionsAndRequiredContent): void;
        toasts: { [toastID: string]: ToastOptionsAndRequiredContent };
      };
      const toast: ToastOptionsAndRequiredContent = {
        id: "id1",
        content: "content",
        timeout: 1000
      };
      vm.setToast(toast);
      expect(vm.toasts["id1"].timeout).toBe(1000);
      vm.updateToast({ id: "id1", options: { timeout: 1000 }, create: false });
      expect(vm.toasts["id1"].timeout).toBe(1001);
    });
    it("creates new toast if create is true", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        updateToast({
          id,
          options,
          create
        }: {
          id: ToastID;
          options: ToastOptionsAndContent;
          create: boolean;
        }): void;
        addToast(params: ToastOptionsAndRequiredContent): void;
        toasts: { [toastID: string]: ToastOptionsAndRequiredContent };
      };
      expect(vm.toasts["id1"]).toBe(undefined);
      vm.updateToast({
        id: "id1",
        options: { content: "content" },
        create: true
      });
      expect(vm.toasts["id1"]).not.toBe(undefined);
      expect(vm.toasts["id1"].content).toBe("content");
    });
    it("ignores if missing toast and not create", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        updateToast({
          id,
          options,
          create
        }: {
          id: ToastID;
          options: ToastOptionsAndContent;
          create: boolean;
        }): void;
        addToast(params: ToastOptionsAndRequiredContent): void;
        toasts: { [toastID: string]: ToastOptionsAndRequiredContent };
      };
      expect(vm.toasts["id1"]).toBe(undefined);
      vm.updateToast({
        id: "id1",
        options: { content: "content" },
        create: false
      });
      expect(vm.toasts["id1"]).toBe(undefined);
    });
  });
  describe("getClasses", () => {
    it("returns classes", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        getClasses(position: POSITION): string[];
      };
      const position = POSITION.BOTTOM_RIGHT;
      expect(vm.getClasses(position)).toEqual([
        `${VT_NAMESPACE}__container`,
        position
      ]);
    });
  });
  describe("toastArray", () => {
    const { containerWrapper } = loadPlugin();
    const vm = (containerWrapper.vm as unknown) as {
      toastArray: ToastOptionsAndRequiredContent[];
      setToast(params: ToastOptionsAndRequiredContent): void;
    };
    expect(vm.toastArray.map(t => t.id)).toEqual([]);
    vm.setToast({ id: "1", content: "abc" });
    expect(vm.toastArray.map(t => t.id)).toEqual(["1"]);
    vm.setToast({ id: "2", content: "def" });
    expect(vm.toastArray.map(t => t.id)).toEqual(["1", "2"]);
  });
  describe("filteredToasts", () => {
    it("filters toasts with default filterToasts", () => {
      const { containerWrapper } = loadPlugin();
      const vm = (containerWrapper.vm as unknown) as {
        filteredToasts: ToastOptionsAndRequiredContent[];
        setToast(params: ToastOptionsAndRequiredContent): void;
      };
      vm.setToast({ id: "1", content: "abc" });
      vm.setToast({ id: "2", content: "def" });
      expect(vm.filteredToasts.map(t => t.id)).toEqual(["1", "2"]);
    });
    it("filters toasts with provided filterToasts", () => {
      const filterToasts = jest.fn(() => []);
      expect(filterToasts).not.toHaveBeenCalled();
      const { containerWrapper } = loadPlugin({ filterToasts });
      const vm = (containerWrapper.vm as unknown) as {
        filteredToasts: ToastOptionsAndRequiredContent[];
        setToast(params: ToastOptionsAndRequiredContent): void;
        toastArray: ToastOptionsAndRequiredContent[];
      };
      expect(filterToasts).toHaveBeenCalledTimes(1);
      expect(filterToasts).toHaveBeenCalledWith([]);
      vm.setToast({ id: "1", content: "abc" });
      vm.setToast({ id: "2", content: "def" });
      expect(filterToasts).toHaveBeenCalledTimes(1);
      expect(vm.filteredToasts).toEqual([]);
      expect(filterToasts).toHaveBeenCalledTimes(2);
      expect(filterToasts).toHaveBeenCalledWith(vm.toastArray);
    });
    it("returns all toasts if filterToasts is undefined", () => {
      const filterToasts = jest.fn(() => []);
      expect(filterToasts).not.toHaveBeenCalled();
      const { containerWrapper } = loadPlugin({ filterToasts });
      const vm = (containerWrapper.vm as unknown) as {
        filteredToasts: ToastOptionsAndRequiredContent[];
        setToast(params: ToastOptionsAndRequiredContent): void;
        toastArray: ToastOptionsAndRequiredContent[];
      };
      expect(filterToasts).toHaveBeenCalledTimes(1);
      expect(filterToasts).toHaveBeenCalledWith([]);
      containerWrapper.vm.$toast.updateDefaults({ filterToasts: undefined });
      vm.setToast({ id: "1", content: "abc" });
      vm.setToast({ id: "2", content: "def" });
      expect(filterToasts).toHaveBeenCalledTimes(1);
      expect(vm.filteredToasts).toEqual(vm.toastArray);
      expect(filterToasts).toHaveBeenCalledTimes(1);
    });
  });
});
