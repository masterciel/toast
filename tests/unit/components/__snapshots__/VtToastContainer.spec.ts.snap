// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VtToastContainer snapshots with classes 1`] = `
<div>
  
  <div>
    <div
      class="Vue-Toastification__container top-left myclass"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container top-center myclass"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container top-right myclass"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container bottom-left myclass"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container bottom-center myclass"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container bottom-right myclass"
    />
  </div>
  
</div>
`;

exports[`VtToastContainer snapshots with default value 1`] = `
<div>
  
  <div>
    <div
      class="Vue-Toastification__container top-left"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container top-center"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container top-right"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container bottom-left"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container bottom-center"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container bottom-right"
    />
  </div>
  
</div>
`;

exports[`VtToastContainer updateDefaults applies new containerClassName 1`] = `
<div>
  
  <div>
    <div
      class="Vue-Toastification__container top-left"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container top-center"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container top-right"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container bottom-left"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container bottom-center"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container bottom-right"
    />
  </div>
  
</div>
`;

exports[`VtToastContainer updateDefaults applies new containerClassName 2`] = `
<div>
  
  <div>
    <div
      class="Vue-Toastification__container top-left my-class"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container top-center my-class"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container top-right my-class"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container bottom-left my-class"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container bottom-center my-class"
    />
  </div>
  <div>
    <div
      class="Vue-Toastification__container bottom-right my-class"
    />
  </div>
  
</div>
`;
